File: lib/matching.py
"""
PSD Officer Rotation Matching Algorithm
Uses Hungarian algorithm (linear_sum_assignment) to optimize officer-position matching
"""

import numpy as np
import pandas as pd
from scipy.optimize import linear_sum_assignment
import json
import sys

def parse_input(input_json):
    """Parse JSON input from API"""
    data = json.loads(input_json)
    
    officers = data['officers']
    positions = data['positions']
    officer_prefs = data['officer_preferences']
    position_prefs = data['position_preferences']
    original_positions = data['original_positions']
    org_prefs = data.get('org_preferences', {})
    
    return officers, positions, officer_prefs, position_prefs, original_positions, org_prefs

def officer_score(officer, position, officer_prefs):
    """Calculate officer's preference score for a position"""
    if officer not in officer_prefs:
        return 0
    
    prefs = officer_prefs[officer]
    if position in prefs:
        # 3 points for 1st choice, 2 for 2nd, 1 for 3rd
        return 4 - (prefs.index(position) + 1)
    return 0

def position_score(position, officer, position_prefs):
    """Calculate position's preference score for an officer"""
    if position not in position_prefs:
        return 0
    
    prefs = position_prefs[position]
    if officer in prefs:
        # 3 points for 1st choice, 2 for 2nd, 1 for 3rd
        return 4 - (prefs.index(officer) + 1)
    return 0

def build_reward_matrix(officers, positions, officer_prefs, position_prefs, original_positions, org_prefs):
    """Build the reward matrix for optimization"""
    num_officers = len(officers)
    num_positions = len(positions)
    
    # Create officer and position ID mappings
    officer_ids = {o: i for i, o in enumerate(officers)}
    position_ids = {p: j for j, p in enumerate(positions)}
    
    # Initialize reward matrix
    reward = np.zeros((num_officers, num_positions))
    
    # Fill reward matrix with preference scores
    for i, officer in enumerate(officers):
        for j, position in enumerate(positions):
            reward[i, j] = (
                officer_score(officer, position, officer_prefs) +
                position_score(position, officer, position_prefs)
            )
    
    # Apply rotation penalty (prevent returning to original position)
    rotation_penalty = -99
    for officer, orig_position in original_positions.items():
        if officer in officer_ids and orig_position in position_ids:
            i = officer_ids[officer]
            j = position_ids[orig_position]
            reward[i, j] += rotation_penalty
    
    # Apply organization preferences
    for key, bonus in org_prefs.items():
        if ',' in key:
            officer, position = key.split(',', 1)
            if officer in officer_ids and position in position_ids:
                i = officer_ids[officer]
                j = position_ids[position]
                reward[i, j] += bonus
    
    return reward, officer_ids, position_ids

def pad_matrix(reward, officers, positions):
    """Pad matrix to handle unequal officers and positions"""
    num_officers = len(officers)
    num_positions = len(positions)
    
    unassigned_penalty = -3
    vacant_penalty = -3
    
    if num_officers < num_positions:
        # More positions than officers - add dummy officers
        num_dummies = num_positions - num_officers
        dummy_rows = np.full((num_dummies, num_positions), vacant_penalty)
        reward_padded = np.vstack([reward, dummy_rows])
        dummy_officers = [f"Vacant_{k+1}" for k in range(num_dummies)]
        all_officers = officers + dummy_officers
        all_positions = positions
        
    elif num_officers > num_positions:
        # More officers than positions - add dummy positions
        num_dummies = num_officers - num_positions
        dummy_cols = np.full((num_officers, num_dummies), unassigned_penalty)
        reward_padded = np.hstack([reward, dummy_cols])
        dummy_positions = [f"Unassigned_{k+1}" for k in range(num_dummies)]
        all_officers = officers
        all_positions = positions + dummy_positions
        
    else:
        # Equal officers and positions
        reward_padded = reward
        all_officers = officers
        all_positions = positions
    
    return reward_padded, all_officers, all_positions

def solve_assignment(reward_padded):
    """Solve the assignment problem using Hungarian algorithm"""
    cost_matrix = -reward_padded  # Convert reward to cost
    row_ind, col_ind = linear_sum_assignment(cost_matrix)
    return row_ind, col_ind

def calculate_metrics(assignments, officer_prefs, position_prefs):
    """Calculate satisfaction metrics"""
    officer_top_counts = {1: 0, 2: 0, 3: 0, 'none': 0}
    position_top_counts = {1: 0, 2: 0, 3: 0, 'none': 0}
    
    for assignment in assignments:
        officer = assignment['officer']
        position = assignment['position']
        
        # Skip dummy entries
        if officer.startswith('Vacant_') or position.startswith('Unassigned_'):
            continue
        
        # Officer satisfaction
        if officer in officer_prefs and position in officer_prefs[officer]:
            rank = officer_prefs[officer].index(position) + 1
            officer_top_counts[rank] += 1
        else:
            officer_top_counts['none'] += 1
        
        # Position satisfaction
        if position in position_prefs and officer in position_prefs[position]:
            rank = position_prefs[position].index(officer) + 1
            position_top_counts[rank] += 1
        else:
            position_top_counts['none'] += 1
    
    return officer_top_counts, position_top_counts

def run_matching(input_json):
    """Main function to run the matching algorithm"""
    try:
        # Parse input
        officers, positions, officer_prefs, position_prefs, original_positions, org_prefs = parse_input(input_json)
        
        # Build reward matrix
        reward, officer_ids, position_ids = build_reward_matrix(
            officers, positions, officer_prefs, position_prefs, original_positions, org_prefs
        )
        
        # Pad matrix if needed
        reward_padded, all_officers, all_positions = pad_matrix(reward, officers, positions)
        
        # Solve assignment problem
        row_ind, col_ind = solve_assignment(reward_padded)
        
        # Extract assignments
        assignments = []
        for i, j in zip(row_ind, col_ind):
            assignments.append({
                'officer': all_officers[i],
                'position': all_positions[j],
                'reward': float(reward_padded[i, j])
            })
        
        # Calculate metrics
        officer_satisfaction, position_satisfaction = calculate_metrics(
            assignments, officer_prefs, position_prefs
        )
        
        # Calculate totals
        total_reward = sum(a['reward'] for a in assignments)
        avg_reward = total_reward / len(assignments)
        
        vacant_count = sum(1 for a in assignments if a['officer'].startswith('Vacant_'))
        unassigned_count = sum(1 for a in assignments if a['position'].startswith('Unassigned_'))
        
        # Return results
        return {
            'success': True,
            'assignments': assignments,
            'metrics': {
                'total_reward': float(total_reward),
                'average_reward': float(avg_reward),
                'officer_satisfaction': officer_satisfaction,
                'position_satisfaction': position_satisfaction,
                'vacant_positions': vacant_count,
                'unassigned_officers': unassigned_count
            }
        }
        
    except Exception as e:
        return {
            'success': False,
            'error': str(e)
        }

if __name__ == '__main__':
    # Read input from stdin
    input_data = sys.stdin.read()
    result = run_matching(input_data)
    print(json.dumps(result))

File: lib/matching.ts
/**
 * PSD Officer Rotation Matching Algorithm
 * TypeScript implementation using Hungarian algorithm
 */

interface MatchingInput {
  officers: string[];
  positions: string[];
  officer_preferences: { [key: string]: string[] };
  position_preferences: { [key: string]: string[] };
  original_positions: { [key: string]: string };
  org_preferences: { [key: string]: number };
}

interface Assignment {
  officer: string;
  position: string;
  reward: number;
}

interface Metrics {
  total_reward: number;
  average_reward: number;
  officer_satisfaction: {
    1: number;
    2: number;
    3: number;
    none: number;
  };
  position_satisfaction: {
    1: number;
    2: number;
    3: number;
    none: number;
  };
  vacant_positions: number;
  unassigned_officers: number;
}

interface MatchingResult {
  success: boolean;
  assignments?: Assignment[];
  metrics?: Metrics;
  error?: string;
}

/**
 * Calculate officer's preference score for a position
 */
function officerScore(officer: string, position: string, officerPrefs: { [key: string]: string[] }): number {
  if (!officerPrefs[officer]) return 0;
  
  const prefs = officerPrefs[officer];
  const index = prefs.indexOf(position);
  
  if (index !== -1) {
    // 3 points for 1st choice, 2 for 2nd, 1 for 3rd
    return 4 - (index + 1);
  }
  return 0;
}

/**
 * Calculate position's preference score for an officer
 */
function positionScore(position: string, officer: string, positionPrefs: { [key: string]: string[] }): number {
  if (!positionPrefs[position]) return 0;
  
  const prefs = positionPrefs[position];
  const index = prefs.indexOf(officer);
  
  if (index !== -1) {
    // 3 points for 1st choice, 2 for 2nd, 1 for 3rd
    return 4 - (index + 1);
  }
  return 0;
}

/**
 * Build the reward matrix for optimization
 */
function buildRewardMatrix(input: MatchingInput): {
  matrix: number[][];
  allOfficers: string[];
  allPositions: string[];
} {
  const { officers, positions, officer_preferences, position_preferences, original_positions, org_preferences } = input;
  
  const numOfficers = officers.length;
  const numPositions = positions.length;
  
  // Initialize reward matrix
  const reward: number[][] = Array(numOfficers).fill(0).map(() => Array(numPositions).fill(0));
  
  // Fill reward matrix with preference scores
  for (let i = 0; i < numOfficers; i++) {
    for (let j = 0; j < numPositions; j++) {
      reward[i][j] = 
        officerScore(officers[i], positions[j], officer_preferences) +
        positionScore(positions[j], officers[i], position_preferences);
    }
  }
  
  // Apply rotation penalty (prevent returning to original position)
  const rotationPenalty = -99;
  for (const [officer, origPosition] of Object.entries(original_positions)) {
    const i = officers.indexOf(officer);
    const j = positions.indexOf(origPosition);
    if (i !== -1 && j !== -1) {
      reward[i][j] += rotationPenalty;
    }
  }
  
  // Apply organization preferences
  for (const [key, bonus] of Object.entries(org_preferences)) {
    if (key.includes(',')) {
      const [officer, position] = key.split(',');
      const i = officers.indexOf(officer);
      const j = positions.indexOf(position);
      if (i !== -1 && j !== -1) {
        reward[i][j] += bonus;
      }
    }
  }
  
  // Pad matrix if needed
  const { matrix: paddedMatrix, allOfficers, allPositions } = padMatrix(reward, officers, positions);
  
  return { matrix: paddedMatrix, allOfficers, allPositions };
}

/**
 * Pad matrix to handle unequal officers and positions
 */
function padMatrix(reward: number[][], officers: string[], positions: string[]): {
  matrix: number[][];
  allOfficers: string[];
  allPositions: string[];
} {
  const numOfficers = officers.length;
  const numPositions = positions.length;
  
  const unassignedPenalty = -3;
  const vacantPenalty = -3;
  
  let paddedMatrix = reward.map(row => [...row]);
  let allOfficers = [...officers];
  let allPositions = [...positions];
  
  if (numOfficers < numPositions) {
    // More positions than officers - add dummy officers
    const numDummies = numPositions - numOfficers;
    for (let k = 0; k < numDummies; k++) {
      paddedMatrix.push(Array(numPositions).fill(vacantPenalty));
      allOfficers.push(`Vacant_${k + 1}`);
    }
  } else if (numOfficers > numPositions) {
    // More officers than positions - add dummy positions
    const numDummies = numOfficers - numPositions;
    for (let i = 0; i < paddedMatrix.length; i++) {
      for (let k = 0; k < numDummies; k++) {
        paddedMatrix[i].push(unassignedPenalty);
      }
    }
    for (let k = 0; k < numDummies; k++) {
      allPositions.push(`Unassigned_${k + 1}`);
    }
  }
  
  return { matrix: paddedMatrix, allOfficers, allPositions };
}

/**
 * Hungarian Algorithm implementation
 * Solves the assignment problem to maximize total reward
 */
function hungarianAlgorithm(costMatrix: number[][]): number[] {
  const n = costMatrix.length;
  const m = costMatrix[0].length;
  
  // Convert reward to cost (negate values)
  const cost = costMatrix.map(row => row.map(val => -val));
  
  // Initialize
  const u = Array(n + 1).fill(0);
  const v = Array(m + 1).fill(0);
  const p = Array(m + 1).fill(0);
  const way = Array(m + 1).fill(0);
  
  for (let i = 1; i <= n; i++) {
    p[0] = i;
    let j0 = 0;
    const minv = Array(m + 1).fill(Infinity);
    const used = Array(m + 1).fill(false);
    
    do {
      used[j0] = true;
      let i0 = p[j0];
      let delta = Infinity;
      let j1 = 0;
      
      for (let j = 1; j <= m; j++) {
        if (!used[j]) {
          const cur = cost[i0 - 1][j - 1] - u[i0] - v[j];
          if (cur < minv[j]) {
            minv[j] = cur;
            way[j] = j0;
          }
          if (minv[j] < delta) {
            delta = minv[j];
            j1 = j;
          }
        }
      }
      
      for (let j = 0; j <= m; j++) {
        if (used[j]) {
          u[p[j]] += delta;
          v[j] -= delta;
        } else {
          minv[j] -= delta;
        }
      }
      
      j0 = j1;
    } while (p[j0] !== 0);
    
    do {
      const j1 = way[j0];
      p[j0] = p[j1];
      j0 = j1;
    } while (j0 !== 0);
  }
  
  const result = Array(n).fill(0);
  for (let j = 1; j <= m; j++) {
    if (p[j] !== 0) {
      result[p[j] - 1] = j - 1;
    }
  }
  
  return result;
}

/**
 * Calculate satisfaction metrics
 */
function calculateMetrics(
  assignments: Assignment[],
  officerPrefs: { [key: string]: string[] },
  positionPrefs: { [key: string]: string[] }
): {
  officerSatisfaction: { 1: number; 2: number; 3: number; none: number };
  positionSatisfaction: { 1: number; 2: number; 3: number; none: number };
} {
  const officerSatisfaction = { 1: 0, 2: 0, 3: 0, none: 0 };
  const positionSatisfaction = { 1: 0, 2: 0, 3: 0, none: 0 };
  
  for (const assignment of assignments) {
    const { officer, position } = assignment;
    
    // Skip dummy entries
    if (officer.startsWith('Vacant_') || position.startsWith('Unassigned_')) {
      continue;
    }
    
    // Officer satisfaction
    if (officerPrefs[officer] && officerPrefs[officer].includes(position)) {
      const rank = (officerPrefs[officer].indexOf(position) + 1) as 1 | 2 | 3;
      officerSatisfaction[rank]++;
    } else {
      officerSatisfaction.none++;
    }
    
    // Position satisfaction
    if (positionPrefs[position] && positionPrefs[position].includes(officer)) {
      const rank = (positionPrefs[position].indexOf(officer) + 1) as 1 | 2 | 3;
      positionSatisfaction[rank]++;
    } else {
      positionSatisfaction.none++;
    }
  }
  
  return { officerSatisfaction, positionSatisfaction };
}

/**
 * Main matching function
 */
export function runMatching(input: MatchingInput): MatchingResult {
  try {
    // Build reward matrix
    const { matrix, allOfficers, allPositions } = buildRewardMatrix(input);
    
    // Solve assignment problem using Hungarian algorithm
    const assignment = hungarianAlgorithm(matrix);
    
    // Extract assignments
    const assignments: Assignment[] = [];
    for (let i = 0; i < assignment.length; i++) {
      const j = assignment[i];
      assignments.push({
        officer: allOfficers[i],
        position: allPositions[j],
        reward: matrix[i][j]
      });
    }
    
    // Calculate metrics
    const { officerSatisfaction, positionSatisfaction } = calculateMetrics(
      assignments,
      input.officer_preferences,
      input.position_preferences
    );
    
    // Calculate totals
    const totalReward = assignments.reduce((sum, a) => sum + a.reward, 0);
    const avgReward = totalReward / assignments.length;
    
    const vacantCount = assignments.filter(a => a.officer.startsWith('Vacant_')).length;
    const unassignedCount = assignments.filter(a => a.position.startsWith('Unassigned_')).length;
    
    return {
      success: true,
      assignments,
      metrics: {
        total_reward: totalReward,
        average_reward: avgReward,
        officer_satisfaction: officerSatisfaction,
        position_satisfaction: positionSatisfaction,
        vacant_positions: vacantCount,
        unassigned_officers: unassignedCount
      }
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error occurred'
    };
  }
}
